### ISSUE TITLE
Question on the methods of Checking and Approving PR's

### ISSUE THREAD

**Issue by ong6** [nus-cs2103-AY2021S2/forum#241]
From what I know, there are 2 main ways one can go about checking and approving PR's:

1. We approve the PR by just looking at the files changed and see if the code looks good
2. We pull the PR locally, test it out and then after testing the PR and code changes related, approve it

The first method is obviously faster but prone to bugs since if no checks are done in the application itself, we might not know if someone's change broke anything. 

The second method is more meticulous but takes up significantly more time especially if the PR is large. 

Is there any recommended way to clearing PR's?

**Comment by kouyk** [comment]
> The first method is obviously faster but prone to bugs since if no checks are done in the application itself, we might not know if someone's change broke anything.

I feel that this point would largely go away with sufficient amount of test code. Since CI has been setup, we should have a decent amount of confidence that if it passes, that should imply that there are no breaking changes unless the PR disabled some test code or broke the test code itself, which is quite evident when looking at the code. Of course all these are on the assumption that there exists sufficiently good amount of test code. And this method is more suitable for individuals familiar with most of the code to know how one component might affect another or at least is well-versed in the particular area being changed.

> The second method is more meticulous but takes up significantly more time especially if the PR is large.

Even in the second method, without test code there will be no sure way to know if the changes broke anything just by manually testing, test code is still necessary to formally verify. In this case, if additional test code is written to verify the PR, we can also make use of the feature of GitHub that allows us to push directly to the branch of the PR and complement the PR (do inform the PR author though). IMO the time taken is actually worthwhile for especially for larger changes to the code base when more tests are written for it in the process. Some examples would include major refactoring or restructuring of the code.

All in all, having relevant test code in place still feels like the key to approving PRs. Without them, I highly doubt it is any different from merging without any review unless the reviewer is an expert with special abilities to simulate the code. Not sure what others think.

**Comment by tlylt** [comment]
> > The first method is obviously faster but prone to bugs since if no checks are done in the application itself, we might not know if someone's change broke anything.
> 
> I feel that this point would largely go away with sufficient amount of test code. Since CI has been setup, we should have a decent amount of confidence that if it passes, that should imply that there are no breaking changes unless the PR disabled some test code or broke the test code itself, which is quite evident when looking at the code. Of course all these are on the assumption that there exists sufficiently good amount of test code. And this method is more suitable for individuals familiar with most of the code to know how one component might affect another or at least is well-versed in the particular area being changed.
> 
> > The second method is more meticulous but takes up significantly more time especially if the PR is large.
> 
> Even in the second method, without test code there will be no sure way to know if the changes broke anything just by manually testing, test code is still necessary to formally verify. In this case, if additional test code is written to verify the PR, we can also make use of the feature of GitHub that allows us to push directly to the branch of the PR and complement the PR (do inform the PR author though). IMO the time taken is actually worthwhile for especially for larger changes to the code base when more tests are written for it in the process. Some examples would include major refactoring or restructuring of the code.
> 
> All in all, having relevant test code in place still feels like the key to approving PRs. Without them, I highly doubt it is any different from merging without any review unless the reviewer is an expert with special abilities to simulate the code. Not sure what others think.

I think you made a good point about having relevant test codes which will prove that PRs that passed the CI checks are, at the very least, worth reviewing. In terms of reviewing PRs, I personally feel that if the changes are non-critical and easily verifiable(for example documentation updates), it is possible to go with the first approach after a cursory check. As for the second approach, it is more suitable when changes are dense and the points mentioned in the PR message are hard to confirm with mere eye power.
That said, I think this brings out the importance of having a proper PR message that conveys the changes, the reasons behind them, and any note-worthy points. If such a message comes along with the PR, it will become a good checklist for PR reviewers to know what to expect.

**Comment by damithc** [comment]
> All in all, having relevant test code in place still feels like the key to approving PRs.

Yup, it does help, especially in catching regressions to other parts of the code. Keep in mind that new tests themselves can be buggy and some aspects may not be covered by tests (e.g., changes to the GUI). In some cases you can ask the developer to post a screencast or screenshots of the new feature in action. At other times, a manual test might be needed to confirm the change is producing the expected behavior change. In the end, the responsibility of testing for bugs is primarily belongs to the developer, not to the PR reviewer.
